File: random_total.swift
import Foundation

func randomTotal(k: Int, n: Int) -> [Int] {
    var test = [Int]()
    var result = [Int]()
    for (var i = 1; i < k; i++) {
        let v = random() % (n - k)
        test.append(v)
    }
    let t = test.sort(<)
    for (var i = 1; i < t.count; i++) {
        result.append(t[i] - t[i-1])
    }
    result.append(t[0] - 0)
    result.append((n - k) - t[t.count-1])
    return result.map( {$0+1} )
}

let k = 4
let n = 9
for (var i = 0; i < 10; i++) {
    let a = randomTotal(k, n: n).sort(<)
    print(a)
    print(a.reduce(0, combine: +))
}

File: random_total.rb
#!/usr/bin/env ruby
# http://programmingpraxis.com/2015/06/12/random-total/

def random_total(k,n)
  test = []; result = []
  (k-1).times { test << rand(n-k) }
  test.sort!
  (1...test.count).each {|i| result << (test[i] - test[i-1])}
  result << test.min - 0
  result << (n - k) - test.max
  result.sort.map {|e| e+1}
end

k = 4; n = 9
10.times { p random_total(k,n).inject(:+) == n }